carla/models/catalog/catalog.py

```diff
@@ -55,73 +59,80 @@ def __init__(
 
         self._model = load_model(model_type, data.name, ext, cache, models_home, **kws)
 
+        self._continuous = data.continous
+        self._categoricals = data.categoricals
+
         self._feature_input_order = feature_input_order
 
+        self._scaler = self.set_scaler(data)
+        self._encoder = self.set_encoder(data)
+
         # Preparing pipeline components
-        self._continuous = data.continous
-        self._categoricals = data.categoricals
-        self._scaler = preprocessing.MinMaxScaler().fit(data.raw[self._continuous])
+        self._use_pipeline = use_pipeline
+        self._pipeline = self.__init_pipeline()
 
-        self._encodings = encoding
+        if encode_normalize_data:
+            data.set_encoded_normalized(self)
 
-    def pipeline(self, df):
-        """
-        Transforms input for prediction into correct form.
-        Only possible for DataFrames without preprocessing steps.
+    def __init_pipeline(self):
+        return [
+            ("scaler", lambda x: scale(self._scaler, self._continuous, x)),
+            ("encoder", lambda x: encode(self._encoder, self._categoricals, x)),
+            ("order", lambda x: order_data(self._feature_input_order, x)),
+        ]
 
-        Recommended to use to keep correct encodings, normalization and input order
+    def get_pipeline_element(self, key):
+        """
+        Returns a specific element of the pipeline
 
         Parameters
         ----------
-        df : pd.DataFrame
-            Contains unnormalized and not encoded data.
+        key : str
+            Element of the pipeline we want to return
 
         Returns
         -------
-        output : pd.DataFrame
-            Prediction input in correct order, normalized and encoded
-
+        Pipeline element
         """
-        output = df.copy()
+        key_idx = list(zip(*self._pipeline))[0].index(key)  # find key in pipeline
+        return self._pipeline[key_idx][1]
 
-        # Normalization
-        output[self._continuous] = self._scaler.transform(output[self._continuous])
-
-        # Encoding
-        output[self._encodings] = 0
-        for encoding in self._encodings:
-            for cat in self._categoricals:
-                if cat in encoding:
-                    value = encoding.split(cat + "_")[-1]
-                    output.loc[output[cat] == value, encoding] = 1
-                    break
-
-        # Get correct order
-        output = output[self._feature_input_order]
+    @property
+    def pipeline(self):
```

Comment:
I saw there is also a sklearn pipeline, I think it does something very similiar to this.

Reply:
Right, I took the sklearn as model and fitted it for our catalog purpose on torch and tensorflow models.
If a user needs a pipeline, he could define its own, or use the sklearn pipeline.